### PR TITLE
STAR-1793 Translate genny ts.ftdc to ftdc

### PR DESCRIPTION
**Pull request by edzh**

### CODE REVIEW

File: t2.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+	for iter.Next() {
+		if err = ctx.Err(); err != nil {
+			return errors.New("operation aborted")
+		}
+		currChunk := iter.Chunk()
+		elems := make([]*birch.Element, 0)
+		var startTime *birch.Element
+
+		timestamp := currChunk.Metrics[0]
```
Comment (julianedwards) [inline t2.go:35]: Add comment clarifying that we expect the 0th index to be the timestamps.

File: t2.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+		for idx, ts := range timestamp.Values {
+			if math.Ceil(float64(ts)/float64(SECOND_MS)) != float64(currentSecond) && endOfSecondIdx > -1 {
+
+				chunk := currChunk
+				if endOfSecondIdx == len(prevChunk.Metrics[0].Values)-1 {
```
Comment (julianedwards) [inline t2.go:41]:
```
idx := i
chunk := currChunk
if currChunk != prevChunk {
        idx = len(prevChunk.Metrics[0].Values) - 1
        chunk = prevChunk
} 
```

Comment (julianedwards) [reply t2.go:41]: I think I would add a comment here as well, clarifying why we need to do this.

File: t2.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+					chunk = prevChunk
+				}
+
+				for _, metric := range chunk.Metrics {
+					switch name := metric.Key(); name {
+					case "ts":
+						startTime = birch.EC.DateTime("start", currentSecond*SECOND_MS)
+					case "counters.n":
+						elems = append(elems, birch.EC.Int64("n", metric.Values[endOfSecondIdx]))
```
Comment (julianedwards) [inline t2.go:50]: Is it possible for the `Values` slices to have different lengths across different metrics within the same chunk? In this case we could panic.

File: t2.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+func TranslateGenny(ctx context.Context, iter *ChunkIterator, output io.Writer, actorOpName string) error {
+	collector := NewStreamingCollector(MAX_SAMPLES, output)
+	var err error
+	defer func() {
+		err = FlushCollector(collector, output)
+	}()
+
+	currentSecond := int64(0)
+	endOfSecondIdx := -1
+	prevChunk := iter.Chunk()
```
Comment (julianedwards) [inline t2.go:25]: I think `iter.Chunk()` will return `nil` until you call `iter.Next()`.

**Review by julianedwards — CHANGES_REQUESTED** [review]

File: t2.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+)
+
+const SECOND_MS int64 = 1000
+const MAX_SAMPLES = 300
+
+func TranslateGenny(ctx context.Context, iter *ChunkIterator, output io.Writer, actorOpName string) error {
+	collector := NewStreamingCollector(MAX_SAMPLES, output)
+	var err error
+	defer func() {
+		err = FlushCollector(collector, output)
```
[the remark this answers is not in the document]
Comment (edzh) [reply t2.go:20]: I added this in order to suppress this lint check: https://spruce.mongodb.com/task/ftdc_lint_lint_ftdc_patch_e0b459ca66b0d5a64c9a74cba49d16cd5316ffd0_60d388493627e03848023d41_21_06_23_19_15_22/logs

do you know a better way?

File: t2.go (hunk 4 above)
Comment (edzh) [reply t2.go:29]: I resolve this by checking if prevChunk is nil, then assign to iter.Chunk() in lines 35-37 with:

```
if prevChunk == nil {
	prevChunk = iter.Chunk()
}
```

File: t2.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+				// If the intended sample to be recorded is in the previous chunk, iterate
+				// through the values of the previous chunk instead. This handles the edge
+				// case where the recorded index is both the last sample of both the chunk
+				// and the whole second window.
+				if currChunk != prevChunk {
+					idx = len(prevChunk.Metrics[0].Values) - 1
+					chunk = prevChunk
+				}
+
+				// create element
```
Comment (julianedwards) [inline t2.go:63]: You can remove this comment.

File: t2.go [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+					default:
+						break
+					}
+				}
+				prevSecond = currSecond
+				prevChunk = currChunk
+			}
+		}
+
+		// create doc
```
Comment (julianedwards) [inline t2.go:93]: You can remove this comment.

File: t2.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (julianedwards) [reply t2.go:25]: I don't think this should be put in a `defer` statement, we probably care about the error `FlushCollector` returns. This should probably be the last line in the function:
```
return errors.Wrap(FlushCollector(collector, output), "flushing collector")
```

File: t2.go (hunk 4 above)
Comment (julianedwards) [reply t2.go:29]: Then I would do `var prevChunk *ftdc.Chunk`

File: t2.go [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+				// create element
+				for _, metric := range chunk.Metrics {
+					switch name := metric.Key(); name {
+					case "ts":
+						startTime = birch.EC.DateTime("start", prevSecond*second_ms)
+					case "counters.n":
+						elems = append(elems, birch.EC.Int64("n", metric.Values[idx]))
+					case "counters.ops":
+						elems = append(elems, birch.EC.Int64("ops", metric.Values[idx]))
```
Comment (julianedwards) [inline t2.go:71]: Would it ever be possible for these `metric.Values` slices to have different lengths? In that case, `idx` could cause an out of index range panic.

**Review by julianedwards — CHANGES_REQUESTED** [review]

File: t2.go (hunk 8 above)
Comment (edzh) [reply t2.go:66]: I'm working on the expectation that a genny workload will output all of the metrics for every event so there shouldn't be an inconsistency between the number of values for each metric

Comment (edzh) [reply t2.go:66]: Raiden just confirmed with me that this is the case.

File: t2.go [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+		if prevChunk == nil {
+			prevChunk = iter.Chunk()
+		}
+		currChunk := iter.Chunk()
+		elems := make([]*birch.Element, 0)
+		var startTime *birch.Element
+
+		// While Metrics can be identified using Metrics[i].Key(),
+		// each metric has a fixed position in the Metrics slice.
+		// The 0th position in Metrics is timestamp
```
Comment (julianedwards) [inline t2.go:40]: Add period.

**Review by julianedwards — APPROVED** [review]
LGTM modulo missing period in comment and passing tests

**Comment by edzh** [comment]
> LGTM modulo missing period in comment and passing tests

https://jira.mongodb.org/browse/STAR-1834 contains the test file I'd like added. I also made changes to the makefile as well if you'd like to quickly review that.
cc @julianedwards

**Pull request merged**